File: 459.RepeatedSubstringPattern/repeated.cc
class Solution {
public:
    bool repeatedSubstringPattern(string s) {
        int end = 1;

        while (true) {
            if (end >= s.length()) {
                return false;
            }

            string pattern = s.substr(0, end);
            int sublen = pattern.length();

            if (s.length() % sublen) {
                end++;
                continue;
            }

            bool flag = true;
            for (int i = end; i < s.length(); i += sublen) {
                if (s.substr(i, end) != pattern) {
                    flag = false;
                    end++;
                    break;
                }
            }

            if (flag) {
                return true;
            }
        }
    }

    bool repeatedSubstringPattern(string s) {
        if (s.length() == 0) {
            return false;
        }
        string s1 = s + s;
        string s2 = s1.substr(1, int(s1.length())-1);
        return s2.find(s) != string::npos;
    }
};

File: 459.RepeatedSubstringPattern/Repeated.java
class Solution {
    public boolean repeatedSubstringPattern(String s) {
        int end = 1;

        while (true) {
            if (end >= s.length()) {
                return false;
            }

            String pattern = s.substring(0, end);
            int sublength = pattern.length();

            if (s.length() % sublength != 0) {
                end++;
                continue;
            }

            boolean flag = true;
            for (int i = end; i < s.length(); i += sublength) {
                if (!s.substring(i, i+sublength).equals(pattern)) {
                    flag = false;
                    end++;
                    break;
                }
            }

            if (flag == true) {
                return true;
            }
        }
    }

    public boolean repeatedSubstringPatternOpt(String s) {
        if (s == null) {
            return false;
        }

        String ss = (s + s)；
        return ss.substring(1, ss.length()-1).indexOf(s) != -1;
    }
}
